lambda-buffers-proto-compat/src/LambdaBuffers/ProtoCompat/ProtoCompat.hs

```diff
@@ -0,0 +1,562 @@
+{-# LANGUAGE KindSignatures, DuplicateRecordFields #-}
+{-# LANGUAGE DeriveGeneric, TypeSynonymInstances #-}
+{-# LANGUAGE DataKinds, OverloadedLabels #-}
+{-# OPTIONS_GHC -Wno-missing-kind-signatures #-}
+{-# OPTIONS_GHC -Wno-missing-deriving-strategies #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+module LambdaBuffers.ProtoCompat.ProtoCompat where
+
+import Data.List.NonEmpty (NonEmpty, nonEmpty)
+import Data.Text (Text)
+import qualified Proto.Compiler as P
+import qualified Proto.Compiler_Fields as P
+import Data.ProtoLens (defMessage)
+import Data.Kind (Type)
+import GHC.Natural (Natural)
+import Control.Lens ( (&), (^.), to, (.~), Getter )
+import Data.Generics.Labels ()
+import GHC.Generics (Generic)
+import Data.Foldable (toList)
+import qualified Data.Text as T
+
+note :: e -> Maybe a -> Either e a
+note e = \case
+  Nothing -> Left e
+  Just a  -> Right a
+
+-- something like this probably exists in lens but i can't find it
+traversing :: (Applicative f,  Traversable t) => (a -> f b) -> Getter (t a) (f (t b))
+traversing f = to $ \ta -> traverse f ta
+
+
+data SourceInfo = SourceInfo {
+  file    :: Text,
+  posFrom :: SourcePosition,
+  posTo   :: SourcePosition
+} deriving (Show, Eq, Ord, Generic)
+
+data SourcePosition = SourcePosition {
+  column :: Int,
+  row    :: Int
+} deriving (Show , Eq, Ord, Generic)
+
+-- for our purposes we really only need a single name type
+data LBName
+  = LBName {name :: Text, sourceInfo :: SourceInfo}
+     deriving (Show, Eq, Ord, Generic)
+
+type TyName     = LBName
+type ConstrName = LBName
+type ModuleName = LBName
+type ArgName    = LBName
+type VarName    = LBName
+type FieldName  = LBName
+type ClassName  = LBName
+
+-- NOTE: Theoretically this could overflow when converting back to Int32, but if you have an arity > 2147483647
+--       then you deserve the resulting errors
+type Kind = Natural
+
+data TyVar = TyVar {
+  varName :: VarName,
+  sourceInfo :: SourceInfo
+} deriving (Show, Eq, Ord, Generic)
+
+data TyInner
+  = TyVarI TyVar
+  | TyAppI TyApp
+  | TyRefI TyRef
+  deriving (Show, Eq, Ord, Generic)
+
+data Ty = Ty {
+  ty :: TyInner,
+  sourceInfo :: SourceInfo
+} deriving (Show, Eq, Ord, Generic)
```

Comment:
I think this is awkward. Why not make a Functor out of Ty like I did in the Frontend? https://github.com/mlabs-haskell/lambda-buffers/blob/main/lambda-buffers-frontend/src/LambdaBuffers/Frontend/Syntax.hs#L26

```
data Ty info
  = TyVar (VarName info) info
  | TyApp (Ty info) [Ty info] info
  | TyRef' (TyRef info) info
  deriving stock (Show, Eq, Ord, Functor, Foldable, Traversable)
```

Reply:
> Wrt to proto-compat:
> 
>     1. Why put that in a separate package? This is Compiler internal right? We should merge this with the compiler lib and expose the module if anyone needs to use it.
> 
>     2. Why use OverloadedLabels and DuplicateRecordFields. Why not just makeLens on all the types?
> 
>     3. Can we please parametrize these types, they will at the very beginning be associated with SourceInfo, but as the processing progresses, kind information will be attached and whatever other annotation might come up.

We discussed 1 and 3 at the meeting but w/r/t 2: Maybe GHC finally got around to fixing this, but as I recall `makeLens` will make things compile a lot slower than Generic Lens will. 

Also, if there's a way to do something without template haskell then _ceteris paribus_ it's a better way to do things. (IMO, maybe you disagree)